File: java/iroha_android/src/main/java/org/hyperledger/iroha/android/offline/OfflineAndroidAttestedDevicePropertiesV2.java
// Copyright 2026 Hyperledger Iroha Contributors
// SPDX-License-Identifier: Apache-2.0

package org.hyperledger.iroha.android.offline;

import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.Objects;

/** Exact bounded Java model of Rust {@code OfflineAndroidAttestedDevicePropertiesV2}. */
public final class OfflineAndroidAttestedDevicePropertiesV2 {
  public static final int VERSION_V2 = 2;
  public static final long OS_VERSION_MAX_V2 = 999_999L;
  public static final int ATTESTED_PROPERTY_MAX_BYTES_V2 = 128;
  public static final int VERIFIED_BOOT_KEY_MAX_BYTES_V2 = 1_024;
  public static final int VERIFIED_BOOT_HASH_BYTES_V2 = 32;
  private static final long U32_MAX = 0xffff_ffffL;

  /** Hardware security boundary authenticated by the Android KeyDescription. */
  public enum SecurityLevel {
    TRUSTED_ENVIRONMENT(0),
    STRONG_BOX(1);

    private final long noritoDiscriminant;

    SecurityLevel(final long noritoDiscriminant) {
      this.noritoDiscriminant = noritoDiscriminant;
    }

    long noritoDiscriminant() {
      return noritoDiscriminant;
    }

    static SecurityLevel fromNoritoDiscriminant(final long value) {
      for (final SecurityLevel candidate : values()) {
        if (candidate.noritoDiscriminant == value) {
          return candidate;
        }
      }
      throw new IllegalArgumentException(
          "unknown Android device security-level discriminant: " + value);
    }
  }

  private final int version;
  private final long attestationVersion;
  private final long keymintVersion;
  private final SecurityLevel securityLevel;
  private final String brand;
  private final String device;
  private final String product;
  private final String manufacturer;
  private final String model;
  private final long osVersion;
  private final long osPatchLevel;
  private final long vendorPatchLevel;
  private final long bootPatchLevel;
  private final byte[] verifiedBootKey;
  private final byte[] verifiedBootHash;

  /**
   * Construct one canonical property snapshot.
   *
   * <p>Empty identity strings and zero version/patch values remain representable because native
   * policy classifies otherwise authenticated but incomplete evidence as drain-only. Call {@link
   * #isCompleteV2()} before authorizing new offline activity.
   */
  public OfflineAndroidAttestedDevicePropertiesV2(
      final int version,
      final long attestationVersion,
      final long keymintVersion,
      final SecurityLevel securityLevel,
      final String brand,
      final String device,
      final String product,
      final String manufacturer,
      final String model,
      final long osVersion,
      final long osPatchLevel,
      final long vendorPatchLevel,
      final long bootPatchLevel,
      final byte[] verifiedBootKey,
      final byte[] verifiedBootHash) {
    if (version != VERSION_V2) {
      throw new IllegalArgumentException(
          "Android attested-device properties version must be exactly 2");
    }
    requireU32(attestationVersion, "attestation_version");
    requireU32(keymintVersion, "keymint_version");
    requireU32(osVersion, "os_version");
    requireU32(osPatchLevel, "os_patch_level");
    requireU32(vendorPatchLevel, "vendor_patch_level");
    requireU32(bootPatchLevel, "boot_patch_level");
    this.version = version;
    this.attestationVersion = attestationVersion;
    this.keymintVersion = keymintVersion;
    this.securityLevel = Objects.requireNonNull(securityLevel, "securityLevel");
    this.brand = requireBoundedProperty(brand, "brand");
    this.device = requireBoundedProperty(device, "device");
    this.product = requireBoundedProperty(product, "product");
    this.manufacturer = requireBoundedProperty(manufacturer, "manufacturer");
    this.model = requireBoundedProperty(model, "model");
    this.osVersion = osVersion;
    this.osPatchLevel = osPatchLevel;
    this.vendorPatchLevel = vendorPatchLevel;
    this.bootPatchLevel = bootPatchLevel;
    this.verifiedBootKey = Objects.requireNonNull(verifiedBootKey, "verifiedBootKey").clone();
    this.verifiedBootHash = Objects.requireNonNull(verifiedBootHash, "verifiedBootHash").clone();
    if (this.verifiedBootKey.length > VERIFIED_BOOT_KEY_MAX_BYTES_V2) {
      throw new IllegalArgumentException(
          "verified_boot_key is outside the V2 protocol bound");
    }
    if (this.verifiedBootHash.length != VERIFIED_BOOT_HASH_BYTES_V2) {
      throw new IllegalArgumentException(
          "verified_boot_hash must contain exactly 32 bytes");
    }
  }

  /** Whether every property required for testnet eligibility is present and canonical. */
  public boolean isCompleteV2() {
    return attestationVersion > 0
        && keymintVersion > 0
        && osVersion > 0
        && osVersion <= OS_VERSION_MAX_V2
        && canonicalPatchMonth(osPatchLevel)
        && canonicalPatchDate(vendorPatchLevel)
        && canonicalPatchDate(bootPatchLevel)
        && verifiedBootKey.length > 0
        && !allZero(verifiedBootHash)
        && !brand.isEmpty()
        && !device.isEmpty()
        && !product.isEmpty()
        && !manufacturer.isEmpty()
        && !model.isEmpty()
        && isAscii(brand)
        && isAscii(device)
        && isAscii(product)
        && isAscii(manufacturer)
        && isAscii(model);
  }

  private static void requireU32(final long value, final String field) {
    if (value < 0 || value > U32_MAX) {
      throw new IllegalArgumentException(field + " must fit in u32");
    }
  }

  private static String requireBoundedProperty(final String value, final String field) {
    Objects.requireNonNull(value, field);
    final byte[] bytes = value.getBytes(StandardCharsets.UTF_8);
    if (bytes.length > ATTESTED_PROPERTY_MAX_BYTES_V2) {
      throw new IllegalArgumentException(field + " exceeds the V2 protocol bound");
    }
    for (int index = 0; index < value.length(); index++) {
      if (Character.isISOControl(value.charAt(index))) {
        throw new IllegalArgumentException(field + " must not contain control characters");
      }
    }
    if (hasBoundaryWhitespace(value)) {
      throw new IllegalArgumentException(field + " must not contain surrounding whitespace");
    }
    return value;
  }

  private static boolean hasBoundaryWhitespace(final String value) {
    if (value.isEmpty()) return false;
    final int first = value.codePointAt(0);
    final int last = value.codePointBefore(value.length());
    return isUnicodeWhitespace(first) || isUnicodeWhitespace(last);
  }

  private static boolean isUnicodeWhitespace(final int codePoint) {
    return Character.isWhitespace(codePoint) || Character.isSpaceChar(codePoint);
  }

  private static boolean canonicalPatchMonth(final long value) {
    if (value < 0 || value > U32_MAX) {
      return false;
    }
    final long year = value / 100;
    final long month = value % 100;
    return year >= 2_000 && year <= 9_999 && month >= 1 && month <= 12;
  }

  private static boolean canonicalPatchDate(final long value) {
    if (value < 0 || value > U32_MAX) {
      return false;
    }
    final long year = value / 10_000;
    final long month = (value / 100) % 100;
    final long day = value % 100;
    if (year < 2_000 || year > 9_999 || month < 1 || month > 12) {
      return false;
    }
    final boolean leap = year % 4 == 0 && (year % 100 != 0 || year % 400 == 0);
    final int maximumDay;
    if (month == 2) {
      maximumDay = leap ? 29 : 28;
    } else if (month == 4 || month == 6 || month == 9 || month == 11) {
      maximumDay = 30;
    } else {
      maximumDay = 31;
    }
    return day >= 1 && day <= maximumDay;
  }

  private static boolean allZero(final byte[] value) {
    int aggregate = 0;
    for (final byte item : value) {
      aggregate |= item;
    }
    return aggregate == 0;
  }

  private static boolean isAscii(final String value) {
    for (int index = 0; index < value.length(); index++) {
      if (value.charAt(index) > 0x7f) return false;
    }
    return true;
  }

  public int version() {
    return version;
  }

  public long attestationVersion() {
    return attestationVersion;
  }

  public long keymintVersion() {
    return keymintVersion;
  }

  public SecurityLevel securityLevel() {
    return securityLevel;
  }

  public String brand() {
    return brand;
  }

  public String device() {
    return device;
  }

  public String product() {
    return product;
  }

  public String manufacturer() {
    return manufacturer;
  }

  public String model() {
    return model;
  }

  public long osVersion() {
    return osVersion;
  }

  public long osPatchLevel() {
    return osPatchLevel;
  }

  public long vendorPatchLevel() {
    return vendorPatchLevel;
  }

  public long bootPatchLevel() {
    return bootPatchLevel;
  }

  public byte[] verifiedBootKey() {
    return verifiedBootKey.clone();
  }

  public byte[] verifiedBootHash() {
    return verifiedBootHash.clone();
  }

  @Override
  public boolean equals(final Object object) {
    if (this == object) {
      return true;
    }
    if (!(object instanceof OfflineAndroidAttestedDevicePropertiesV2 other)) {
      return false;
    }
    return version == other.version
        && attestationVersion == other.attestationVersion
        && keymintVersion == other.keymintVersion
        && osVersion == other.osVersion
        && osPatchLevel == other.osPatchLevel
        && vendorPatchLevel == other.vendorPatchLevel
        && bootPatchLevel == other.bootPatchLevel
        && securityLevel == other.securityLevel
        && brand.equals(other.brand)
        && device.equals(other.device)
        && product.equals(other.product)
        && manufacturer.equals(other.manufacturer)
        && model.equals(other.model)
        && Arrays.equals(verifiedBootKey, other.verifiedBootKey)
        && Arrays.equals(verifiedBootHash, other.verifiedBootHash);
  }

  @Override
  public int hashCode() {
    int result =
        Objects.hash(
            version,
            attestationVersion,
            keymintVersion,
            securityLevel,
            brand,
            device,
            product,
            manufacturer,
            model,
            osVersion,
            osPatchLevel,
            vendorPatchLevel,
            bootPatchLevel);
    result = 31 * result + Arrays.hashCode(verifiedBootKey);
    result = 31 * result + Arrays.hashCode(verifiedBootHash);
    return result;
  }
}

File: kotlin/core-jvm/src/main/java/org/hyperledger/iroha/sdk/offline/OfflineAndroidAttestedDevicePropertiesV2.kt
// Copyright 2026 Hyperledger Iroha Contributors
// SPDX-License-Identifier: Apache-2.0

package org.hyperledger.iroha.sdk.offline

/** Hardware security boundary authenticated by an Android KeyDescription. */
enum class OfflineAndroidDeviceSecurityLevelV2(
    internal val noritoDiscriminant: Long,
) {
    /** Key isolated by the device trusted execution environment. */
    TRUSTED_ENVIRONMENT(0),

    /** Key isolated by a discrete StrongBox secure element. */
    STRONG_BOX(1),
    ;

    internal companion object {
        fun fromNoritoDiscriminant(value: Long): OfflineAndroidDeviceSecurityLevelV2 =
            entries.firstOrNull { it.noritoDiscriminant == value }
                ?: throw IllegalArgumentException(
                    "unknown Android device security-level discriminant: $value",
                )
    }
}

/**
 * Exact bounded SDK model of Rust `OfflineAndroidAttestedDevicePropertiesV2`.
 *
 * Empty identity strings and zero version/patch values remain representable because native
 * policy classifies otherwise authenticated but incomplete evidence as drain-only. Call
 * [isCompleteV2] before using the properties to authorize new offline activity.
 */
class OfflineAndroidAttestedDevicePropertiesV2(
    val version: Int,
    val attestationVersion: Long,
    val keymintVersion: Long,
    val securityLevel: OfflineAndroidDeviceSecurityLevelV2,
    val brand: String,
    val device: String,
    val product: String,
    val manufacturer: String,
    val model: String,
    val osVersion: Long,
    val osPatchLevel: Long,
    val vendorPatchLevel: Long,
    val bootPatchLevel: Long,
    verifiedBootKey: ByteArray,
    verifiedBootHash: ByteArray,
) {
    private val _verifiedBootKey = verifiedBootKey.copyOf()
    private val _verifiedBootHash = verifiedBootHash.copyOf()

    val verifiedBootKey: ByteArray
        get() = _verifiedBootKey.copyOf()

    val verifiedBootHash: ByteArray
        get() = _verifiedBootHash.copyOf()

    init {
        require(version == VERSION_V2) {
            "Android attested-device properties version must be exactly 2"
        }
        require(attestationVersion in 0..U32_MAX) {
            "attestation_version must fit in u32"
        }
        require(keymintVersion in 0..U32_MAX) {
            "keymint_version must fit in u32"
        }
        require(osVersion in 0..U32_MAX) {
            "os_version must fit in u32"
        }
        require(osPatchLevel in 0..U32_MAX) { "os_patch_level must fit in u32" }
        require(vendorPatchLevel in 0..U32_MAX) { "vendor_patch_level must fit in u32" }
        require(bootPatchLevel in 0..U32_MAX) { "boot_patch_level must fit in u32" }
        for ((field, value) in listOf(
            "brand" to brand,
            "device" to device,
            "product" to product,
            "manufacturer" to manufacturer,
            "model" to model,
        )) {
            requireBoundedProperty(value, field)
        }
        require(_verifiedBootKey.size <= VERIFIED_BOOT_KEY_MAX_BYTES_V2) {
            "verified_boot_key is outside the V2 protocol bound"
        }
        require(_verifiedBootHash.size == VERIFIED_BOOT_HASH_BYTES_V2) {
            "verified_boot_hash must contain exactly 32 bytes"
        }
    }

    /** Whether every property required for testnet eligibility is present and canonical. */
    fun isCompleteV2(): Boolean =
        attestationVersion > 0 &&
            keymintVersion > 0 &&
            osVersion in 1..OS_VERSION_MAX_V2 &&
            canonicalPatchMonth(osPatchLevel) &&
            canonicalPatchDate(vendorPatchLevel) &&
            canonicalPatchDate(bootPatchLevel) &&
            _verifiedBootKey.isNotEmpty() &&
            _verifiedBootHash.any { it.toInt() != 0 } &&
            listOf(brand, device, product, manufacturer, model).all { value ->
                value.isNotEmpty() && value.all { character -> character.code <= 0x7f }
            }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is OfflineAndroidAttestedDevicePropertiesV2) return false
        return version == other.version &&
            attestationVersion == other.attestationVersion &&
            keymintVersion == other.keymintVersion &&
            securityLevel == other.securityLevel &&
            brand == other.brand &&
            device == other.device &&
            product == other.product &&
            manufacturer == other.manufacturer &&
            model == other.model &&
            osVersion == other.osVersion &&
            osPatchLevel == other.osPatchLevel &&
            vendorPatchLevel == other.vendorPatchLevel &&
            bootPatchLevel == other.bootPatchLevel &&
            _verifiedBootKey.contentEquals(other._verifiedBootKey) &&
            _verifiedBootHash.contentEquals(other._verifiedBootHash)
    }

    override fun hashCode(): Int {
        var result = version
        for (value in listOf<Any>(
            attestationVersion,
            keymintVersion,
            securityLevel,
            brand,
            device,
            product,
            manufacturer,
            model,
            osVersion,
            osPatchLevel,
            vendorPatchLevel,
            bootPatchLevel,
        )) {
            result = 31 * result + value.hashCode()
        }
        result = 31 * result + _verifiedBootKey.contentHashCode()
        result = 31 * result + _verifiedBootHash.contentHashCode()
        return result
    }

    companion object {
        const val VERSION_V2: Int = 2
        const val OS_VERSION_MAX_V2: Long = 999_999
        const val ATTESTED_PROPERTY_MAX_BYTES_V2: Int = 128
        const val VERIFIED_BOOT_KEY_MAX_BYTES_V2: Int = 1_024
        const val VERIFIED_BOOT_HASH_BYTES_V2: Int = 32
        private const val U32_MAX: Long = 0xffff_ffffL

        private fun requireBoundedProperty(value: String, field: String) {
            require(value.toByteArray(Charsets.UTF_8).size <= ATTESTED_PROPERTY_MAX_BYTES_V2) {
                "$field exceeds the V2 protocol bound"
            }
            require(value.none(Char::isISOControl)) {
                "$field must not contain control characters"
            }
            require(value == value.trim()) { "$field must not contain surrounding whitespace" }
        }

        private fun canonicalPatchMonth(value: Long): Boolean {
            if (value !in 0..U32_MAX) return false
            val year = value / 100
            val month = value % 100
            return year in 2_000..9_999 && month in 1..12
        }

        private fun canonicalPatchDate(value: Long): Boolean {
            if (value !in 0..U32_MAX) return false
            val year = value / 10_000
            val month = (value / 100) % 100
            val day = value % 100
            if (year !in 2_000..9_999 || month !in 1..12) return false
            val leap = year % 4 == 0L && (year % 100 != 0L || year % 400 == 0L)
            val maximumDay = when (month) {
                2L -> if (leap) 29 else 28
                4L, 6L, 9L, 11L -> 30
                else -> 31
            }
            return day in 1..maximumDay
        }
    }
}
